[SYCL] Split 'double type' test into two tests, 1) <test>.cpp; 2) <test>_aspect-fp64.cpp for llvm-test-suite

myler opened the pull request:
'double' type are not native supported in Intel Gen11/12 and next Gens GPUs in the future, to disable particularly big number of important tests is not reasonable, so split 'double' part into two separate test file.

- split double tests in AtomicRef directory
- split double tests in Basic directory
- split double tests in DeprecatedFeatures directory
- split double tests in DeviceLib directory
- split double tests in ESIMD directory
- split double tests in GroupAlgorithm directory
- split double tests in InlineAsm directory
- split double tests in KernelParams directory
- split double tests in Regression directory
- split double tests in SpecConstants directory
- split double tests in SubGroup directory
- split double tests in USM directory
- add `REQUIRES:aspect-fp64` to restrict some 'double' type tests "_double.cpp", "_fp64.cpp", "_fp_extra.cpp", etc.

myler committed:
split double tests in AtomicRef directory

myler committed:
split double tests in Basic directory

myler committed:
split double tests in DeprecatedFeatures directory

myler committed:
split double tests in DeviceLib directory

myler committed:
split double tests in ESIMD directory

myler committed:
split double tests in GroupAlgorithm directory

myler committed:
split double tests in InlineAsm directory

myler committed:
split double tests in KernelParams directory

myler committed:
split double tests in Regression directory

myler committed:
split double tests in SpecConstants directory

myler committed:
split double tests in SubGroup directory

myler committed:
split double tests in USM directory

myler committed:
remove deprecated namespace cl::

myler committed:
fix clang-format issue

myler committed:
fix clang-format issue 2

myler committed:
fix 3-way merge conflicts in union_kernel_param.cpp

myler commented:
Failures are unrelated:

[2022-08-17T07:35:47.081Z] Failed Tests (1):
[2022-08-17T07:35:47.081Z]   SYCL :: InvokeSimd/invoke_simd_conv.cpp

[2022-08-17T07:41:39.229Z] Unexpectedly Passed Tests (2):
[2022-08-17T07:41:39.229Z]   SYCL :: SubGroup/reduce_spirv13.cpp
[2022-08-17T07:41:39.229Z]   SYCL :: SubGroup/reduce_spirv13_fp64.cpp

v-klochkov commented on SYCL/ESIMD/api/bin_and_cmp_ops_heavy_aspect-fp64.cpp:
```diff
@@ -0,0 +1,55 @@
+  using CmpOp = esimd_test::CmpOp;
+  auto cmp_ops = esimd_test::CmpOps;
+  passed &= test<short, double, 7, CmpOp, VSf, IDf>(cmp_ops, q);
+  passed &= test<double, half, 7, CmpOp, VSf, IDf>(cmp_ops, q);
```
If aspect::fp64 is checked for 'double', then shouldn't 'aspect::fp16 be also checked for 'half' ?

myler replied:
We have many aspects in sycl spec 2020, but we don't plan to separate files for all of them. AFAIK, Intel newer GPUs will no longer support fp64 (double) natively, so that's why we need to split 'aspect::fp64'.

Tagging @pvchupin who suggests this change instead of disabling all important tests.

v-klochkov replied:
I am Ok with not splitting/separating fp16 for ESIMD test. 
For more general and non-ESIMD tests (if there are such in this PR), using fp16 aspect check would make more sense, IMO. Thank you.

myler replied:
Will do 'aspect::fp16 to check half' in another PR later.

v-klochkov replied:
Ok, Thank you!

aelovikov-intel commented on SYCL/USM/copy.cpp:
```diff
@@ -12,83 +12,30 @@
-}
-
-template <typename T> void runTests(queue q, T val, alloc kind1, alloc kind2) {
-  bool dev_dst1 = (kind1 == alloc::device);
```
Why isn't it enough to make early returns from this function, without creating an extra .cpp file? Same for other places.

myler replied:
Not sure you are asking for `dev_dst1 = (kind1 == alloc::device);`?
I guess you mean returns from this function if the device `!has(sycl::aspect::fp64)`?
The JIT compiler raises unsupported 'double' type error in compiling no matter we have `if/else` statement to return early. We have to split them into two TUs.

aelovikov-intel replied:
How do customers solve this? Imagine I have fast double algorithm and slow float algorithm, and I want to do host-based dispatch to run one based on the current device. Are you saying this is impossible?

pvchupin replied:
Do we have compiler implementation lagging behind? Probably not everything from SYCL2020 in this regard is implemented yet.

Tagging @gmlueck

gmlueck replied:
Yes, this is a bug in DPC++, and we have had customer complaints about it.  Applications should not be required to put their "double" kernels in a separate source file.  The SYCL 2020 specification requires that this work even when there is a single source file.  The proposed solution is the design that is outlined in [OptionalDeviceFeatures](https://github.com/intel/llvm/blob/sycl/sycl/doc/design/OptionalDeviceFeatures.md).

pvchupin replied:
@AlexeySachkov do we plan to address it soon? Otherwise we should split to keep testing specific underlying features properly.

aelovikov-intel replied:
I still think splitting like this increases maintainability burden greatly. I'd rather see something like:

```
// test.cpp
// RUN: ... -DENABLE_FP64=false
constexpr bool EnableFP64 = ENABLE_FP64;

void test() {
   // non-fp64-case
   do_smth();

   if constexpr (EnableFP64)
     do_smth64();
}

// test-fp64.cpp
; RUN: ... -DENABLE_FP64=true
#include <test-fp.cpp>
```

myler replied:
@aelovikov-intel, that's good solution!
The preprocess step is able to strip out those fp64 code, and no complain is raised from GPU compiler `error: Double type is not supported on this platform.` any more.
I will update this change in 24 files `<test>_aspect-fp64.cpp` very soon.

aelovikov-intel replied:
Thanks!

myler committed:
To avoid splitting increases maintainability burden greatly, use `-DENABLE_FP64` macro to split 'double' type code as following:
```
// test.cpp
// RUN: ... -DENABLE_FP64=false
constexpr bool EnableFP64 = ENABLE_FP64;

void test() {
   // non-fp64-case
   do_smth();

   if constexpr (EnableFP64)
     do_smth64();
}

// test-fp64.cpp
; RUN: ... -DENABLE_FP64=true
\#include <test-fp.cpp>
```

myler committed:
fix 3-way merge issue

myler committed:
Merge branch 'intel' into split_double_tests

myler committed:
fix clang-format issue

myler committed:
fix clang-format issue

myler committed:
fix 3-way merge issue in subgroup_info test, and split 'double' code into info_aspect-fp64.cpp

myler commented:
Unrelated failures:
```
[2022-08-23T09:33:09.407Z] Failed Tests (5):
[2022-08-23T09:33:09.407Z]   SYCL :: DiscardEvents/discard_events_accessors.cpp
[2022-08-23T09:33:09.407Z]   SYCL :: DiscardEvents/discard_events_using_assert.cpp
[2022-08-23T09:33:09.407Z]   SYCL :: DiscardEvents/discard_events_using_assert_ndebug.cpp
[2022-08-23T09:33:09.407Z]   SYCL :: DiscardEvents/discard_events_usm.cpp
[2022-08-23T09:33:09.407Z]   SYCL :: DiscardEvents/discard_events_usm_ooo_queue.cpp
```

aelovikov-intel commented on SYCL/Basic/buffer/buffer.cpp:
```diff
@@ -532,174 +539,190 @@ int main() {
-          double_vector[i] != 7.5) {
-        assert(false && "Data was not copied back");
-        return 1;
+      if (bool_vector[i] != true || int_vector[i] != 3) {
+#ifdef ENABLE_FP64
+        if (bool_vector[i] != true || int_vector[i] != 3 ||
+            double_vector[i] != 7.5) {
+#endif
```
Maybe just me, but this is hard to read. Also, I think there is a functional change here. Consider the case when bool/int are ok and double is wrong. We won't even take the first if so the double check will not be done at all. I'd change like this:

```
  bool Passed = true;
  Passed &= (bool_vector[i] == true);
  Passed &= (int_vector[i] == 3);
  #ifdef ENABLE_FP64
  Passed &= (double_vector[i] == 7.5);
  #endif
```

aelovikov-intel commented on SYCL/Basic/buffer/buffer.cpp:
```diff
@@ -532,174 +539,190 @@ int main() {
-  // set_write_back
-  {
-    std::vector<int> data1(10, -1);
+    // Check that data is not copied back after canceling write-back using
```
Unrelated change?

myler replied:
I just applied the auto generated patch by the clang-format-check testing.

aelovikov-intel replied:
It looks like previously line 528 scope was closed at line 573 and it isn't closed there anymore. Hard to track that in the web-browser though, so I might be wrong.

aelovikov-intel replied:
I don't think this has been addressed...

myler replied:
Sorry for the mistake, fixed the in-correct brackets `{`.

aelovikov-intel commented on SYCL/ESIMD/api/bin_and_cmp_ops_heavy_aspect-fp64.cpp:
```diff
@@ -0,0 +1,27 @@
+//===----------------------------------------------------------------------===//
+// REQUIRES: aspect-fp64, gpu
+// UNSUPPORTED: cuda || hip
+// TODO: esimd_emulator fails due to unimplemented 'half' type
+// XFAIL: esimd_emulator
+// RUN: %clangxx -fsycl -DENABLE_FP64 %s -o %t.out
+// RUN: %GPU_RUN_PLACEHOLDER %t.out
+
+// Tests various binary operations applied to simd objects.
+
+// TODO
+// Arithmetic operations behaviour depends on Gen's control regiter's rounding
+// mode, which is RTNE by default:
+//    cr0.5:4 is 00b = Round to Nearest or Even (RTNE)
+// For half this leads to divergence between Gen and host (emulated) results
+// larger than certain threshold. Might need to tune the cr0 once this feature
+// is available in ESIMD.
+//
```
I don't think we need to duplicate the title comments in fp64 versions. I'd change all of the them to a generic

```
Enable FP64 part of <orig_test_name.cpp>. To be removed once DPC++
supports optional device features and the code could be enabled
unconditionally without causing failures in speculative compilation
of the kernels.
```

myler replied:
Applied this to all new added fp64 sources.

aelovikov-intel replied:
But what about this one on lines 21-30 ?

aelovikov-intel commented on SYCL/ESIMD/ext_math.cpp:
```diff
@@ -467,17 +467,21 @@ int main(void) {
     Pass &= testSYCL<float, 32>(Q);
   }
+#ifdef ENABLE_FP64
   Pass &= testESIMDSqrtIEEE<float, 16>(Q);
```
Why is this float guarded?

myler replied:
Re-enable those IEEE-conformant single precision checks.

aelovikov-intel commented on SYCL/ESIMD/regression/Inputs/dgetrf.hpp:
```diff
@@ -18,10 +18,16 @@
 #include <sycl/ext/intel/esimd.hpp>
 #include <sycl/sycl.hpp>
 
+#ifdef ENABLE_FP64
+typedef double fptype;
+#else
+typedef float fptype;
+#endif
```
```suggestion
#ifdef ENABLE_FP64
using fptype = double;
#else
using fptype = float;
#endif
```

aelovikov-intel commented on SYCL/ESIMD/regression/Inputs/dgetrf.hpp:
```diff
@@ -47,7 +53,7 @@ using namespace sycl;
 using namespace sycl::ext::intel::esimd;
 
-ESIMD_PRIVATE ESIMD_REGISTER(192) simd<double, 3 * 32 * 4> GRF;
+ESIMD_PRIVATE ESIMD_REGISTER(256) simd<fptype, 3 * 32 * 4> GRF;
```
@intel/dpcpp-esimd-reviewers , is this change (192->256) ok?

myler replied:
Revert the change back to `192`, this is an internal requirement, should not merge to public repo.

aelovikov-intel commented on SYCL/ESIMD/regression/dgetrf_8x8.cpp:
```diff
@@ -20,10 +20,16 @@
 #include <sycl/ext/intel/esimd.hpp>
 #include <sycl/sycl.hpp>
 
+#ifdef ENABLE_FP64
+typedef double fptype;
+#else
+typedef float fptype;
+#endif
+
```
Please change to use C++ style type aliases `using fptype = ....`.

aelovikov-intel commented on SYCL/InlineAsm/asm_float_add.cpp:
```diff
@@ -9,7 +9,7 @@
 #include <vector>
 
-using dataType = sycl::cl_double;
+using dataType = sycl::cl_float;
```
Why this change? Is there a second part of it somewhere in this patch?

aelovikov-intel commented on SYCL/InlineAsm/asm_float_imm_arg.cpp:
```diff
@@ -9,8 +9,8 @@
 
-constexpr double IMM_ARGUMENT = 0.5;
-using dataType = sycl::cl_double;
+constexpr float IMM_ARGUMENT = 0.5;
```
Likewise - why this change?

aelovikov-intel commented on SYCL/KernelParams/union_kernel_param.cpp:
```diff
@@ -9,31 +9,37 @@
 #include <cstdio>
 #include <sycl/sycl.hpp>
 
+#ifdef ENABLE_FP64
+typedef double fptype;
+#else
+typedef float fptype;
+#endif
```
Please use C++ style type alias.

aelovikov-intel commented on SYCL/SpecConstants/2020/handler-api.cpp:
```diff
@@ -23,8 +23,10 @@
 constexpr sycl::specialization_id<int> int_id2(2);
-constexpr sycl::specialization_id<double> double_id(3.14);
 constexpr sycl::specialization_id<custom_type> custom_type_id;
+#ifdef ENABLE_FP64
+constexpr sycl::specialization_id<double> double_id(3.14);
+#endif
```
I don't think we really need to move it to after `custom_type_id` (imagine we'll remove `#ifdef` without moving it back), but I can live with it.

Same below.

myler replied:
Moved `fp64` code block before `custom_type_id`

myler committed:
Added common comment in <origin test name>_aspect-fp64.cpp, fixed typo issue and changed to use C++ style type aliases.

myler committed:
fix clang-format issue

myler committed:
fix unmatched brackets

myler commented:
Unrelated failure:
```
[2022-08-24T04:03:12.640Z] Failed Tests (1):
[2022-08-24T04:03:12.640Z]   SYCL :: DeprecatedFeatures/buffer_interop.cpp
```

aelovikov-intel commented on SYCL/ESIMD/api/saturation_smoke_aspect-fp64.cpp:
```diff
@@ -0,0 +1,23 @@
+//==---- saturation_smoke_aspect-fp64.cpp  - DPC++ ESIMD on-device test
+//-----==//
```
This looks wrong.

myler replied:
This change is still required by the clang-format-check patch.

aelovikov-intel replied:
Shorten the "---" parts and it will fit one line.

aelovikov-intel commented on SYCL/InlineAsm/asm_float_add_aspect-fp64.cpp:
```diff
@@ -0,0 +1,11 @@
+// Enable FP64 part of <asm_float_add.cpp>. To be removed once DPC++
+// supports optional device features and the code could be enabled
+// unconditionally without causing failures in speculative compilation
+// of the kernels.
```
I don't think this is accurate. Removal of the ENABLE_FP64 would result in dropping one of the cases in the main file, not executing them both.

myler committed:
use function template to run the test for multiple data types in asm_float_add and re-enable IEEE-conformant single precision check in ESIMD/ext_math.cpp

myler committed:
shorten the comment line < 80 characters.

myler committed:
fix clang-format issue

aelovikov-intel commented on SYCL/ESIMD/regression/dgetrf_8x8_aspect-fp64.cpp:
```diff
@@ -0,0 +1,18 @@
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+// Enable FP64 part of <dgetrf_8x8.cpp>. To be removed once DPC++
+// supports optional device features and the code could be enabled
+// unconditionally without causing failures in speculative compilation
+// of the kernels.
```
I think I've mentioned that before, maybe for another test. This is incorrect as doubles aren't filtered out in the included sources (to the best of my understanding). Instead, we have a selection between float/double and they cannot co-exist at the same time.

myler replied:
Fixed in `dgetrf.cpp` by using function template.

aelovikov-intel commented on SYCL/InlineAsm/asm_float_imm_arg_aspect-fp64.cpp:
```diff
@@ -0,0 +1,11 @@
+// Enable FP64 part of <asm_float_imm_arg.cpp>. To be removed once DPC++
+// supports optional device features and the code could be enabled
+// unconditionally without causing failures in speculative compilation
+// of the kernels.
```
Same as above.

myler replied:
Fixed in `asm_float_imm_arg.cpp` by using function template.

aelovikov-intel commented on SYCL/SubGroup/generic-shuffle_aspect-fp64.cpp:
```diff
@@ -0,0 +1,21 @@
+// RUN: %GPU_RUN_PLACEHOLDER %t.out
+// RUN: %ACC_RUN_PLACEHOLDER %t.out
+//
+//==-- generic_shuffle_aspect-fp64.cpp - SYCL sub_group generic shuffle test *-
+// C++ -*--==//
```
This looks wrong. Can you please fix such dangling formatting throughout the whole PR?

myler committed:
fix all dangling formatting, use function template to check float/double.

myler committed:
fix clang-format issue

myler committed:
typo fix in dgetrf_8x8.cpp

myler committed:
fix clang-format issue

myler committed:
1) add REQUIRES:aspect-fp64 to restrict some 'double' type tests running on Gen12 or late
2) fix typo issue in dgetrf_8x8.cpp
3) split test replicate_smoke.cpp
4) reslve conflict in SYCL/ESIMD/api/bin_and_cmp_ops_heavy.cpp

myler committed:
Merge branch 'intel' into split_double_tests

myler committed:
use '-device gen12lp' for aot_mixed.cpp

myler commented:
Related failure:
```
[2022-08-30T07:31:19.596Z] Failed Tests (1):
[2022-08-30T07:31:19.596Z]   SYCL :: ESIMD/aot_mixed.cpp
```
This error `LLVM ERROR: vc::translateBuild: unsupported platform` cannot be reproduced.
```
Running on Intel(R) Iris(R) Xe Graphics
Running ESIMD kernel...
TEST Passed
```


Unrelated failures:
```
[2022-08-30T08:08:17.655Z] Failed Tests (4):
[2022-08-30T08:08:17.655Z]   SYCL :: ESIMD/api/replicate_smoke.cpp
[2022-08-30T08:08:17.655Z]   SYCL :: ESIMD/api/simd_view_select_2d_int.cpp
[2022-08-30T08:08:17.655Z]   SYCL :: ESIMD/ext_math.cpp
[2022-08-30T08:08:17.655Z]   SYCL :: ESIMD/matrix_transpose_glb.cpp
```

v-klochkov reviewed (dismissed):
Ok for ESIMD. Thank you.

I also have a non-blocking comment regarding the naming.
IMO, it is better to stick to underscores and not mix them with '-'.
E.g. "functions_select_2d_core_aspect_fp64.cpp" looks much better to me than "functions_select_2d_core_aspect-fp64.cpp"

AlexeySachkov reviewed (dismissed):
LGTM for spec constants

myler committed:
rename aspect-fp64 to aspect_fp64 to stick to underscores and not mix then with '-'

myler committed:
also to rename the file markup

myler commented:
> Ok for ESIMD. Thank you.
> 
> I also have a non-blocking comment regarding the naming. IMO, it is better to stick to underscores and not mix them with '-'. E.g. "functions_select_2d_core_aspect_fp64.cpp" looks much better to me than "functions_select_2d_core_aspect-fp64.cpp"

Renamed these new added file with only underscores.

myler commented:
Hi @Pennycook,
Seems other subdirectories are already approved, but may I ask you to review changes in `SubGroup` and `GroupAlgorithm`, seems you are the code owner of them.

myler commented:
Related failure:
```
[2022-08-31T02:01:02.817Z] Failed Tests (1):
[2022-08-31T02:01:02.817Z]   SYCL :: ESIMD/aot_mixed.cpp
```
This error `LLVM ERROR: vc::translateBuild: unsupported platform` cannot be reproduced.
```
Running on Intel(R) Iris(R) Xe Graphics
Running ESIMD kernel...
TEST Passed
```
Unrelated failures:
```
[2022-08-31T02:34:16.124Z] Failed Tests (4):
[2022-08-31T02:34:16.124Z]   SYCL :: ESIMD/api/replicate_smoke.cpp
[2022-08-31T02:34:16.124Z]   SYCL :: ESIMD/api/simd_view_select_2d_int.cpp
[2022-08-31T02:34:16.124Z]   SYCL :: ESIMD/ext_math.cpp
[2022-08-31T02:34:16.124Z]   SYCL :: ESIMD/matrix_transpose_glb.cpp
```

romanovvlad commented:
A possible alternative solution is here: https://github.com/intel/llvm-test-suite/pull/1190/files

myler committed:
Merge branch 'intel' into split_double_tests

myler commented:
Failures are known and unrelated.
Hi @v-klochkov 
According to the https://github.com/intel/llvm-test-suite/blob/intel/CONTRIBUTING.md#merge, need maintainer to merge the PR. Could you help me on this?

myler committed:
Merge remote-tracking branch 'upstream/intel' into split_double_tests

v-klochkov commented:
> Failures are known and unrelated. Hi @v-klochkov According to the https://github.com/intel/llvm-test-suite/blob/intel/CONTRIBUTING.md#merge, need maintainer to merge the PR. Could you help me on this?

I did not merge this yet because of the comment from @romanovvlad (https://github.com/intel/llvm-test-suite/pull/1150#issuecomment-1232874469) and did not have clear understanding if the discussion here https://github.com/intel/llvm-test-suite/pull/1190 has got some solution/agreement.

That alternative approach used in https://github.com/intel/llvm-test-suite/pull/1190 has sympathy/votes from at least 3 people. 
And I would add my vote to there as well because device.has is specially designed and convenient way to run-or-skip hw specific features.

I gave my approval for this PR, but I'd rather let @romanovvlad to merge this patch if Vlad's comment was non-blocking

AlexeySachkov commented:
Personally, I would vote for device code split approach, because it allows us to avoid duplicating coverage for non-optional types. I can help with creating PRs similar to one that @romanovvlad published (the first one to go is #1241)

pvchupin commented:
@v-klochkov, @myler , @romanovvlad , @AlexeySachkov 
I assume we decided to proceed with alternative approach partially done at #1241 , #1190?

@aelovikov-intel, can you please take a look at remaining ones on Windows? We currently have ~30 fails and it's really unsafe to merge anything else until these fixed.

pvchupin commented:
@myler, should we close this PR in favor of alternatives in progress?

myler commented:
> @myler, should we close this PR in favor of alternatives in progress?

Hi @pvchupin 

This solution tends to change the testing framework `REQURIES: aspect-fp64` and macros `#ifdef ENABLE_FP64`, not depends on SYCL standard `aspect::fp64` and feature `-fsycl-device-code-split=per_kernel`

#1190 tends to use SYCL style solution to handle `double` type tests.

If the final decision is to use `if (device.has(aspect::fp64))` + `-fsycl-device-code-split=per_kernel` to handle `double` type tests, I agree to close this one. But #1190 did not fix all tests, we may have at least 26 tests contain `double` type code.

gmlueck commented:
> If the final decision is to use if (device.has(aspect::fp64)) + -fsycl-device-code-split=per_kernel to handle double type tests, I agree to close this one. But https://github.com/intel/llvm-test-suite/pull/1190 did not fix all tests, we may have at least 26 tests contain double type code.

That seems like a better solution to me.  Once we implement the [OptionalDeviceFeatures](https://github.com/intel/llvm/blob/sycl/sycl/doc/design/OptionalDeviceFeatures.md) design, we can remove the "-fsycl-device-code-split=per_kernel", which should be relatively easy.

Are you saying  there are 26 test failures that will not be fixed with this technique?  Do we know what the cause of those failures is?

myler commented:
Close this PR as we have another alternative.
